code/ui/blocks/src/blocks/Canvas.tsx

```diff
@@ -54,13 +124,88 @@ const usePreviewProps = (
   };
 };
 
-export const Canvas: FC<CanvasProps> = (props) => {
+export const Canvas: FC<CanvasProps & DeprecatedCanvasProps> = (props) => {
   const docsContext = useContext(DocsContext);
   const sourceContext = useContext(SourceContext);
-  const { isLoading, previewProps } = usePreviewProps(props, docsContext, sourceContext);
-  const { children } = props;
+  const { children, of, source } = props;
+  const { isLoading, previewProps } = useDeprecatedPreviewProps(props, docsContext, sourceContext);
+  let story;
+  let sourceProps;
+  try {
+    /**
+     * useOf and useSourceProps will throw if they can't find the story, in the scenario where
+     * the doc is unattached (no primary story) and 'of' is undefined.
+     * That scenario is valid in the deprecated API, where children is used as story refs rather than 'of'.
+     * So if children is passed we allow the error to be swallowed and we'll use them instead.
+     */
+    ({ story } = useOf(of || 'story', ['story']));
+    sourceProps = useSourceProps({ ...source, of }, docsContext, sourceContext);
+  } catch (error) {
+    if (!children) {
+      throw error;
+    }
+  }
 
-  if (isLoading) return <PreviewSkeleton />;
+  if (props.withSource) {
+    deprecate(dedent`Setting source state with \`withSource\` is deprecated, please use \`sourceState\` with 'hidden', 'shown' or 'none' instead. 
+    
+    Please refer to the migration guide: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#canvas-block'
+    `);
+  }
+  if (props.mdxSource) {
+    deprecate(dedent`Setting source code with \`mdxSource\` is deprecated, please use source={{code: '...'}} instead. 
+    
+    Please refer to the migration guide: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#canvas-block'
+    `);
+  }
+  if (props.isColumn !== undefined || props.columns !== undefined) {
+    deprecate(dedent`\`isColumn\` and \`columns\` props are deprecated as the Canvas block now only supports showing a single story. 
+    
+    Please refer to the migration guide: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#canvas-block'
+    `);
+  }
+  if (children) {
+    deprecate(dedent`Passing children to Canvas is deprecated, please use the \`of\` prop instead to reference a story. 
+    
+    Please refer to the migration guide: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#canvas-block'
+  `);
+    return isLoading ? (
+      <PreviewSkeleton />
+    ) : (
+      <PurePreview {...previewProps}>{children}</PurePreview>
```

Comment:
I was only focusing on fixing the layout (and other bugs) in the new API, and keep the deprecated use "broken". If we feel it's important to fix the old usage as well, I'll do an attempt.

Reply:
Well the old usage won't work any more (I don't think?) because the `Story` block doesn't pass the `parameters` prop down to the `Story` component. Or maybe it does though?

Reply:
It should work, layout never had anything to do with Story, it was always only an thing on Canvas/Preview, I don't know why the API was to assign it to the Story.

Either way, I've added a couple of internal stories to test the deprecated API and that it works.